pkg/port/grpc.go

```diff
@@ -1,198 +1,194 @@
+func (s *Server) DeleteBridgePort(_ context.Context, in *pb.DeleteBridgePortRequest) (*emptypb.Empty, error) {
 	// check input correctness
 	if err := s.validateDeleteBridgePortRequest(in); err != nil {
+		log.Printf("DeleteBridgePort(): validation failure: %v", err)
 		return nil, err
 	}
 	// fetch object from the database
-	iface := new(pb.BridgePort)
-	ok, err := s.store.Get(in.Name, iface)
+	_, err := s.getBridgePort(in.Name)
 	if err != nil {
-		fmt.Printf("Failed to interact with store: %v", err)
-		return nil, err
-	}
-	if !ok {
-		if in.AllowMissing {
-			return &emptypb.Empty{}, nil
+		if err != infradb.ErrKeyNotFound {
+			log.Printf("Failed to interact with store: %v", err)
+			return nil, err
 		}
-		err := status.Errorf(codes.NotFound, "unable to find key %s", in.Name)
-		return nil, err
-	}
-	// configure netlink
-	if err := s.netlinkDeleteBridgePort(ctx, iface); err != nil {
-		return nil, err
+		if !in.AllowMissing {
+			err = status.Errorf(codes.NotFound, "unable to find key %s", in.Name)
+			log.Printf("DeleteBridgePort(): BridgePort with id %v: Not Found %v", in.Name, err)
+			return nil, err
+		}
+		return &emptypb.Empty{}, nil
 	}
-	// remove from the Database
-	delete(s.ListHelper, iface.Name)
-	err = s.store.Delete(iface.Name)
-	if err != nil {
+
+	if err := s.deleteBridgePort(in.Name); err != nil {
+		log.Printf("DeleteBridgePort(): BridgePort with id %v, Delete Bridge Port from DB failure: %v", in.Name, err)
 		return nil, err
 	}
 	return &emptypb.Empty{}, nil
 }
 
 // UpdateBridgePort updates an Nvme Subsystem
-func (s *Server) UpdateBridgePort(ctx context.Context, in *pb.UpdateBridgePortRequest) (*pb.BridgePort, error) {
+func (s *Server) UpdateBridgePort(_ context.Context, in *pb.UpdateBridgePortRequest) (*pb.BridgePort, error) {
 	// check input correctness
 	if err := s.validateUpdateBridgePortRequest(in); err != nil {
+		log.Printf("UpdateBridgePort(): validation failure: %v", err)
 		return nil, err
 	}
 	// fetch object from the
-	port := new(pb.BridgePort)
-	ok, err := s.store.Get(in.BridgePort.Name, port)
+	bpObj, err := s.getBridgePort(in.BridgePort.Name)
 	if err != nil {
-		fmt.Printf("Failed to interact with store: %v", err)
-		return nil, err
-	}
-	if !ok {
-		// TODO: introduce "in.AllowMissing" field. In case "true", create a new resource, don't return error
-		err := status.Errorf(codes.NotFound, "unable to find key %s", in.BridgePort.Name)
-		return nil, err
+		if err != infradb.ErrKeyNotFound {
+			log.Printf("UpdateBridgePort(): Failed to interact with store: %v", err)
+			return nil, err
+		}
+		if !in.AllowMissing {
+			err = status.Errorf(codes.NotFound, "unable to find key %s", in.BridgePort.Name)
+			log.Printf("UpdateBridgePort(): BridgePort with id %v: Not Found %v", in.BridgePort.Name, err)
+			return nil, err
+		}
+
+		log.Printf("UpdateBridgePort(): Bridge Port with id %v is not found so it will be created", in.BridgePort.Name)
```

Comment:
Should the Update really do an auto-create if the port is not found or should it just return not found?  The update may not have all the parameters needed since it typically has a update map of parameters being changed.

Reply:
It depends. According to the google AIP guidelines if we have the AllowMissing to true then it will create the object otherwise it will return an error. Check line 103. 

Also the createBridgePort function internally validates also the received Spec so if a parameter is missing that needs to be there it will return error.

The Update functions are not yet tested. We have only the implementation there but is untested for the update case. We will test that in the future but right now is not first priority to have a fully working update. We have tested the Create, Delete, Get and List but not the update yet

Reply:
Ok, so, if the check in line 103 doesn't execute it falls through and then calls the "createBridgePort (function in line 27) and handles the processing there based on the parameters passed in and doesn't worry about the update mask settings.

My concern is that if only the update mask parameters are provided in the update request, then somewhere the defaults need to be set for the interface being created.

Reply:
Actually in order to allow the creation of an object using the field mask we first validate that the filed mask includes all the mandatory parameters. We do not really have any defaults. If the mandatory parameters are not set then we will return an error to the user. 

But the update call is an untested call so far. I would like to test that in the future and revisit the matter. However I think that we are pretty safe with the current implementation that we have now